### PR TITLE
Wei Ning's CIS 194 Week-1 homework

### PR DESCRIPTION
**Pull request by powergun**
Hi Shine,
Please take a look at my homework. 
My apology for late submission. 
I'm not too sure if the coding style is correct (I'm using the default setting of stylish haskell).
Please run `stack test` to execute the test cases.

Thank you very much! I really enjoyed the first week of learning.

### CODE REVIEW

File: cis194/week1/wei-ning/app/Main.hs [hunk 1]
```diff
@@ -0,0 +1,2 @@
+module Main where
+main = undefined
```
Comment (cht8687) [inline cis194/week1/wei-ning/app/Main.hs]: nice undefined 😄

File: cis194/week1/wei-ning/src/Exercise1.hs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  assert "ex1.1 given num, expect list" $ [1..9] == toDigits 123456789
+  assert "ex1.1 given num, expect reverse list" $ reverse [1..9] == toDigitsRev 123456789
+
+toDigits :: Integer -> [Integer]
+toDigitsRev :: Integer -> [Integer]
+
+toDigitsRev x | x <= 0    = []
+              | x > 0 && x < 10 = [x]
+              | otherwise = (x `mod` 10 ): toDigitsRev (x `div` 10)
+toDigits = reverse . toDigitsRev
```
Comment (cht8687) [inline cis194/week1/wei-ning/src/Exercise1.hs]: @shineli1984 is it ok in real Haskell project to mix test with implementation?

File: cis194/week1/wei-ning/src/Assert.hs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+module Assert (assert) where
+
+assert :: String -> Bool -> IO ()
+assert msg expr =
+  if expr
+  then putStrLn (msg ++ " : passed")
+  else error    (msg ++ " : failed")
```
Comment (cht8687) [inline cis194/week1/wei-ning/src/Assert.hs]: @shineli1984 in Haskell, is if then else encouraged?

File: cis194/week1/wei-ning/src/Exercise3.hs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+runTests =
+  assert "ex1.3 expect sum of digits" $ 22 == sumDigits [16, 7, 12, 5]
+
+sumDigits :: [Integer] -> Integer
+sumDigits = foldr sumTwoNums 0
+
+sumTwoNums :: Integer -> Integer -> Integer
+sumTwoNums x y =
+  if x > 9
+  then (x `div` 10) + sumTwoNums (x `mod` 10) y
```
Comment (shineli1984) [inline cis194/week1/wei-ning/src/Exercise3.hs]: Can we reuse function from exercise 1?

File: cis194/week1/wei-ning/src/Exercise5.hs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+runTests =
+  assert "ex1.5 expect total num of moves" $ 2^15 - 1 == length (hanoi 15 "a" "b" "c")
+
+type Peg = String
+type Move = (Peg, Peg)
+
+hanoi :: Integer -> Peg -> Peg -> Peg -> [Move]
+hanoi 1 from to _    = [(from, to)]
+hanoi n from to temp =
+  hanoi (n - 1) from temp to ++
```
Comment (shineli1984) [inline cis194/week1/wei-ning/src/Exercise5.hs]: In Haskell, We usually align symbols. This means ++ will be at the beginning of a new line aligning with = sign above.

File: cis194/week1/wei-ning/src/Assert.hs (hunk 3 above)
Comment (zehua) [reply cis194/week1/wei-ning/src/Assert.hs]: how about?
```
assert msg True  = putStrLn (msg ++ " : passed")
assert msg False = error    (msg ++ " : failed")
```

File: cis194/week1/wei-ning/src/Exercise4.hs [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+import           Exercise2 (doubleEveryOther)
+import           Exercise3 (sumDigits)
+
+runTests :: IO ()
+runTests = do
+  assert "ex1.4 expect valid credit card num" $ validate 4012888888881881
+  assert "ex1.4 expect invalid credit card num" $ not (validate 4012888888881882)
+
+validate :: Integer -> Bool
+validate = (== 0) . (`mod` 10) . sumDigits . doubleEveryOther . toDigitsRev
```
Comment (cht8687) [inline cis194/week1/wei-ning/src/Exercise4.hs]: seems `mod` 10 is not correct. as well as toDigitsRev. is Rev necessary?
here is the description of the task:
```
Calculate the remainder when the sum is divided by 10. For the
above example, the remainder would be 8.
```

Comment (cht8687) [reply cis194/week1/wei-ning/src/Exercise4.hs]: oh, sorry actually, this describes `mod` 10

File: cis194/week1/wei-ning/src/Exercise1.hs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  assert "ex1.1 given 0, expect empty list" $ null (toDigitsRev 0)
+  assert "ex1.1 given num, expect list" $ [1..9] == toDigits 123456789
+  assert "ex1.1 given num, expect reverse list" $ reverse [1..9] == toDigitsRev 123456789
+
+toDigits :: Integer -> [Integer]
+toDigitsRev :: Integer -> [Integer]
+
+toDigitsRev x | x <= 0    = []
+              | x > 0 && x < 10 = [x]
+              | otherwise = (x `mod` 10 ): toDigitsRev (x `div` 10)
```
Comment (zehua) [inline cis194/week1/wei-ning/src/Exercise1.hs]: space before `:` /nit

File: cis194/week1/wei-ning/src/Exercise3.hs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+runTests :: IO ()
+runTests =
+  assert "ex1.3 expect sum of digits" $ 22 == sumDigits [16, 7, 12, 5]
+
+sumDigits :: [Integer] -> Integer
+sumDigits = foldr sumTwoNums 0
+
+sumTwoNums :: Integer -> Integer -> Integer
+sumTwoNums x y =
+  if x > 9
```
Comment (zehua) [inline cis194/week1/wei-ning/src/Exercise3.hs]: consider using `guard`

File: cis194/week1/wei-ning/src/Exercise1.hs (hunk 2 above)
Comment (powergun) [reply cis194/week1/wei-ning/src/Exercise1.hs]: Thanks. I will fix this.

File: cis194/week1/wei-ning/src/Assert.hs (hunk 3 above)
Comment (powergun) [reply cis194/week1/wei-ning/src/Assert.hs]: Yeah can do.
"Prefer guards over if-then-else where possible." - haskell style guide https://kowainik.github.io/posts/2019-02-06-style-guide#if-then-else-clauses

File: cis194/week1/wei-ning/src/Exercise1.hs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  assert "ex1.1 given -1, expect empty list" $ null (toDigits (-1))
+  assert "ex1.1 given 0, expect empty list" $ null (toDigitsRev 0)
+  assert "ex1.1 given num, expect list" $ [1..9] == toDigits 123456789
+  assert "ex1.1 given num, expect reverse list" $ reverse [1..9] == toDigitsRev 123456789
+
+toDigits :: Integer -> [Integer]
+toDigitsRev :: Integer -> [Integer]
+
+toDigitsRev x | x <= 0    = []
+              | x > 0 && x < 10 = [x]
```
[the remark this answers is not in the document]
Comment (powergun) [reply cis194/week1/wei-ning/src/Exercise1.hs]: Yes that's right! Good catch

File: cis194/week1/wei-ning/src/Exercise5.hs (hunk 5 above)
Comment (powergun) [reply cis194/week1/wei-ning/src/Exercise5.hs]: will do. Thx

File: cis194/week1/wei-ning/src/Exercise3.hs (hunk 8 above)
Comment (powergun) [reply cis194/week1/wei-ning/src/Exercise3.hs]: will do thx!

File: cis194/week1/wei-ning/src/Exercise3.hs (hunk 4 above)
Comment (powergun) [reply cis194/week1/wei-ning/src/Exercise3.hs]: > Can we reuse function from exercise 1?

Yes I realise it is doing exactly what `toDigits` does.

**Comment by powergun** [comment]
Thanks for reviewing @shineli1984 @cht8687 @zehua

**Pull request merged**